integration_test.go

```diff
@@ -0,0 +1,231 @@
+package endless
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+	"sync"
+	"syscall"
+	"testing"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	IntegrationServerFeatureToggle = "GO_INTEGRATION_TEST_SERVER"
+	IntegrationServerPort          = "4242"
+	IntegrationServerPath          = "/test"
+)
+
+var (
+	serverpids = make(map[int]struct{})
+)
+
+// TestIntegrationServer isn't a real test - instead it is used as a
+// test server binary so that we can test signal handling and the subsequent
+// shutdown/restarts from a separate binary from the test itself.
+// This test is only run when an environment variable feature toggle
+// (named by IntegrationServerFeatureToggle) is set by the calling test
+func TestIntegrationServer(t *testing.T) {
+	if os.Getenv(IntegrationServerFeatureToggle) != "1" {
+		t.Skip("Skipping as do not need to stand up integration test server in this run")
+	}
+
+	// Create new HTTP server
+	mux1 := mux.NewRouter()
+
+	mux1.HandleFunc(IntegrationServerPath, integrationTestServerHandler).
+		Methods("GET")
+
+	srv := NewServer("tcp", "localhost:"+IntegrationServerPort, mux1)
+	srv.Debug = true
+
+	// This is used to signal back to the calling test that the server has started up
+	fmt.Println("PID:", os.Getpid())
+	if err := srv.ListenAndServe(); err != nil {
+		fmt.Println(err)
+	}
+}
+
+// handler used by the integration test server
+func integrationTestServerHandler(w http.ResponseWriter, r *http.Request) {
+	duration, err := time.ParseDuration(r.FormValue("duration"))
+	if err != nil {
+		http.Error(w, err.Error(), 400)
+		return
+	}
+	time.Sleep(duration)
+	fmt.Fprintf(
+		w, "Slept for %2v seconds. Response from pid %d.\n",
+		duration.Seconds(),
+		os.Getpid(),
+	)
+}
+
+// createIntegrationServer will re-run the test binary that is currently running,
+// but only execute the "TestIntegrationServer" test, which will stand up an
+// 'endless' server which can be used to test signal handling
+func createIntegrationServerTest() *exec.Cmd {
+	cmd := exec.Command(os.Args[0], "-test.run=TestIntegrationServer")
+
+	// Set the feature toggle for the integration test server
+	cmd.Env = []string{IntegrationServerFeatureToggle + "=1"}
+
+	return cmd
+}
+
+func TestIntegrationRestart(t *testing.T) {
+
+	intServer := createIntegrationServerTest()
+	serverMessages := make(chan string) // channel to receive notifications from the integration server
```

Comment:
Quite a bit of unneeded comments here, the code speaks for itself, so this only adds review overhead.

We typically only comment where the behaviour isn't obvious.